contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -73,153 +76,211 @@ contract LidoOracle is ILidoOracle, AragonApp {
+        QUORUM_CALLBACK_POSITION.setStorageUint256(uint256(_addr));
+        emit QuorumCallbackSet(_addr);
     }
 
     /**
      * @notice An oracle committee member reports data from the ETH 2.0 side
-     * @param _epochId BeaconChain epoch id
+     * @param _epochId Beacon Chain epoch id
      * @param _beaconBalance Balance in wei on the ETH 2.0 side
      * @param _beaconValidators Number of validators visible on this epoch
      */
     function reportBeacon(uint256 _epochId, uint128 _beaconBalance, uint128 _beaconValidators) external {
-        (uint256 startEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
-        require(_epochId >= startEpoch, "EPOCH_IS_TOO_OLD");
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        (uint256 currentEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
+        require(_epochId >= currentEpoch, "EPOCH_IS_TOO_OLD");
         require(_epochId <= endEpoch, "EPOCH_HAS_NOT_YET_BEGUN");
+        require(_epochId.mod(beaconSpec.epochsPerFrame) == 0, "MUST_BE_FIRST_EPOCH_FRAME");
 
-        address member = msg.sender;
-        uint256 index = _getMemberId(member);
-        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        // If reported epoch has advanced, clear the last unsuccessful reporting
+        if (_epochId > currentEpoch) _clearReportingAndAdvanceTo(_epochId);
 
-        // check & set contribution flag
-        uint256 bitMask = gatheredEpochData[_epochId].reportsBitMask;
+        // Make sure the oracle is from members list and has not yet voted
+        uint256 index = _getMemberId(msg.sender);
+        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        uint256 bitMask = REPORTS_BITMASK_POSITION.getStorageUint256();
         require(!bitMask.getBit(index), "ALREADY_SUBMITTED");
+        REPORTS_BITMASK_POSITION.setStorageUint256(bitMask.setBit(index, true));
 
-        LAST_REPORTED_EPOCH_ID_POSITION.setStorageUint256(_epochId);
-
-        gatheredEpochData[_epochId].reportsBitMask = bitMask.setBit(index, true);
-
-        Report memory currentReport = Report(_beaconBalance, _beaconValidators);
-        gatheredEpochData[_epochId].reports[index] = currentReport;
-
+        // Push this report to the matching kind
+        uint256 i = 0;
+        Report memory report = Report(_beaconBalance, _beaconValidators);
+        uint256 reportRaw = reportToUint256(report);
+
+        // so iterate on all report kinds we alredy have, not more then oracle members maximum
+        while (i < gatheredReportKinds.length && reportToUint256(gatheredReportKinds[i].report) != reportRaw) ++i;
+        if (i < gatheredReportKinds.length) {
+            ++gatheredReportKinds[i].count;
+        } else {
+            gatheredReportKinds.push(ReportKind(report, 1));
+        }
+        emit BeaconReported(_epochId, _beaconBalance, _beaconValidators, msg.sender);
         _tryPush(_epochId);
```

Comment:
Every time the oracle reports the value, the `_tryPush` method is called. I suggest saving gas and not calling the method if there are fewer reports than quorum in the current epoch. What do you think?

Reply:
how do you you suppose to check that? use popcnt?

Reply:
Fix in https://github.com/lidofinance/lido-dao/pull/311. After the fix, no need to `_tryPush` every time.